core/arch/riscv64/riscv64.asm

```diff
@@ -300,30 +300,84 @@ GLOBAL_LABEL(dynamorio_app_take_over:)
  *                       int sysnum,                      // A1
  *                       ptr_uint_t sys_arg1/param_base,  // A2
  *                       ptr_uint_t sys_arg2,             // A3
- *                       bool exitproc,                   // A4
- *                       (2 more args that are ignored: Mac-only))
+ *                       bool exitproc)                   // A4
  *
  * See decl in arch_exports.h for description.
  */
         DECLARE_FUNC(cleanup_and_terminate)
 GLOBAL_LABEL(cleanup_and_terminate:)
-/* FIXME i#3544: Not implemented */
-        jal       GLOBAL_REF(unexpected_return)
+        /* move argument registers to callee saved registers */
+        mv       s0, ARG1  /* dcontext ptr size */
+        mv       s1, ARG2  /* sysnum */
+        mv       s2, ARG3  /* sys_arg1 */
+        mv       s3, ARG4  /* sys_arg2 */
+        mv       s4, ARG5  /* exitproc */
+                          /* s5 reserved for dstack ptr */
+                          /* s6 reserved for syscall ptr */
+
+        li       a1, 1
+        /* inc exiting_thread_count to avoid being killed once off all_threads list */
+        la       a0, GLOBAL_REF(exiting_thread_count)
+        amoadd.w zero, a1, (a0)
+
+        /* save dcontext->dstack for freeing later and set dcontext->is_exiting */
+        sw       a1, is_exiting_OFFSET(s0) /* dcontext->is_exiting = 1 */
+        CALLC1(GLOBAL_REF(is_currently_on_dstack), s0)
+        bnez     a0, cat_save_dstack
+        li       s5, 0
+        j        cat_done_saving_dstack
+cat_save_dstack:
+        ld       s5, dstack_OFFSET(s0)
+cat_done_saving_dstack:
+        CALLC0(GLOBAL_REF(get_cleanup_and_terminate_global_do_syscall_entry))
+        mv       s6, a0
+        beqz     s4, cat_thread_only
+        CALLC0(GLOBAL_REF(dynamo_process_exit))
+        j        cat_no_thread
+cat_thread_only:
+        CALLC0(GLOBAL_REF(dynamo_thread_exit))
+cat_no_thread:
+        /* switch to d_r_initstack for cleanup of dstack */
+        la       s7, GLOBAL_REF(initstack_mutex)
+cat_spin:
+        li       a0, 1
+        amoswap.w a0, a0, (s7)
+        bnez     a0, cat_spin
```

Comment:
Is there no yield or pause instruction?  If so, please add; if not, please add a comment saying we're missing what we'd like to do and what we do on other arches.

Reply:
We don't have a YIELD-like hint instruction like what's available on Aarch64, so we use a plain spin lock here.

Is this comment good?

Reply:
Sure.